File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/parser/CustomInfoRegistry.java
package me.bristermitten.mittenlib.annotations.parser;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import com.squareup.javapoet.TypeName;

import javax.lang.model.type.TypeMirror;
import java.util.Optional;

public abstract class CustomInfoRegistry<T> {
    private final Multimap<TypeName, T> infoMultimap = HashMultimap.create();

    public void register(TypeName clazz, T info) {
        infoMultimap.put(clazz, info);
    }


    public Optional<T> getCustomInfo(TypeMirror propertyType) {
        var fromMap = infoMultimap.get(TypeName.get(propertyType));

        if (fromMap.isEmpty()) {
            return Optional.empty();
        }
        if (fromMap.size() > 1) {
            throw new IllegalArgumentException("Not sure how to handle multiple yet");
        }

        return Optional.of(fromMap.iterator().next());
    }

}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/parser/CustomInfoRegistry.scala
package me.bristermitten.mittenlib.annotations.parser

import com.google.common.collect.{HashMultimap, Multimap}
import com.palantir.javapoet.TypeName
import io.toolisticon.aptk.tools.MessagerUtils
import javax.lang.model.`type`.TypeMirror
import scala.jdk.CollectionConverters.*

abstract class CustomInfoRegistry[T]:
  private val infoMultimap: Multimap[TypeName, T] = HashMultimap.create()

  def register(clazz: TypeName, info: T): Unit =
    infoMultimap.put(clazz, info)

  def getCustomInfo(propertyType: TypeMirror): Option[T] =
    val fromMap = infoMultimap.get(TypeName.get(propertyType))
    if (fromMap.isEmpty) {
      None
    } else if (fromMap.size() > 1) {
      MessagerUtils.error(
        null: javax.lang.model.element.Element,
        s"Multiple custom registrations found for type $propertyType"
      )
      Some(fromMap.iterator().next())
    } else {
      Some(fromMap.iterator().next())
    }

  def allInfos: Map[TypeName, List[T]] =
    infoMultimap
      .asMap()
      .asScala
      .map { case (k, v) =>
        k -> v.asScala.toList
      }
      .toMap
